### PR TITLE
Auto gather

### PR DESCRIPTION
**Pull request by EdgesFTW**
Runalyzer script can create an in-memory database when provided multiple ungathered sqlite files.

### CODE REVIEW

**Review by matthiasdiener — COMMENTED** [review]
This looks good! Two suggestions for the moment:

- Please add a test in `.github/workflows/ci.yaml`
- Please add this somewhere in the docs

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -88,6 +88,52 @@ jobs:
             ## Check that t_init is within [0, 20], and that there is only one value
             runalyzer -m summary.sqlite -c 'assert all([0 < r[0] < 20 for r in q("select $t_init.max")])'
             runalyzer -m summary.sqlite -c 'assert len([r[0] for r in q("select $t_init.max")]) == 1'
+        - name: Run and test example without gathering
+          run: |
+            set -x
+            python examples/log.py
```
Comment (matthiasdiener) [inline .github/workflows/ci.yaml:94]: Can you test this without rerunning the example, and comparing the results of the autogathered values with the gathered file?

File: bin/runalyzer [hunk 2]
```diff
@@ -12,8 +12,10 @@ def main() -> None:
             help="disable mangling of SQL queries")
     parser.add_argument("-c", "--commands", action="store",
             help="commands to execute")
-    parser.add_argument("dbfile", help="database file to read")
-    parser.add_argument("scriptfile", nargs="?", help="script file to read")
+    parser.add_argument("dbfiles", nargs="+", type=str,
+                        help="database file to read")
```
Comment (matthiasdiener) [inline bin/runalyzer:16]:
```suggestion
                        help="database file(s) to read")
```

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (matthiasdiener) [reply .github/workflows/ci.yaml:94]: Also, maybe you can gather and autogather multiple runs together and compare the output.

File: logpyle/runalyzer.py [hunk 3]
```diff
@@ -454,11 +455,57 @@ def my_sprintf(format: str, arg: str) -> str:
...
+        # list of run features as {name: sql_type}
+        fg = FeatureGatherer(False, None)
+        features, dbname_to_run_id = scan(fg, infiles)
+
+        fmap = make_name_map("")
+        qmap = make_name_map("")
+
+        connection = gather_multi_file(":memory:", infiles, fmap, qmap, fg, features,
+                                 dbname_to_run_id, True)
+        assert type(connection) is sqlite3.Connection
```
Comment (matthiasdiener) [inline logpyle/runalyzer.py:501]: Maybe:

```suggestion
        assert connection
```

File: logpyle/runalyzer_gather.py [hunk 4]
```diff
@@ -339,4 +341,8 @@ def transfer_data_table_multi(db_conn: Connection, tbl_name: str,
     pb.finished()  # type: ignore[no-untyped-call]
 
     db_conn.commit()
-    db_conn.close()
+    if return_conn:
+        return db_conn
+    else:
+        db_conn.close()
+        return None
```
Comment (matthiasdiener) [inline logpyle/runalyzer_gather.py:348]: maybe always return the conn and expect the caller to close it?

File: doc/log.rst [hunk 5]
```diff
@@ -20,28 +20,32 @@ given as "watches" are also periodically logged to the console.
 To analyze this data, we need some tools that are distributed as part of this
 package.
 
-To prepare this data set for analysis, run ``runalyzer-gather`` as a script::
+You may use ``runalyzer`` to analyze the summary file::
+
+    $ runalyzer summary.sqlite
```
Comment (matthiasdiener) [inline doc/log.rst:25]:
```suggestion
You may use ``runalyzer`` to analyze the log file::

    $ runalyzer log.sqlite
```

File: logpyle/runalyzer_gather.py (hunk 4 above)
Comment (EdgesFTW) [reply logpyle/runalyzer_gather.py]: Currently the RunDB class closes the connection in its destructor

File: logpyle/runalyzer.py [hunk 6]
```diff
@@ -454,11 +458,56 @@ def my_sprintf(format: str, arg: str) -> str:
...
+        if len(res) == 1:
+            gathered = True
+
+    if gathered:
+        # gathered files should only have one file
+        if len(filenames) > 1:
+            raise Exception("Runalyzing multiple gathered files is not supported!!!")
+
+        return sqlite3.connect(filenames[0])
+    else:
```
Comment (matthiasdiener) [inline logpyle/runalyzer.py:488]: No need for the `else` I think.

File: logpyle/runalyzer.py [hunk 7]
```diff
@@ -43,6 +44,9 @@ def __init__(self, db: Connection, interactive: bool) -> None:
         self.interactive = interactive
         self.rank_agg_tables: Set[Tuple[str, Callable[..., Any]]] = set()
 
+    def __del__(self) -> None:
+        self.db.close()
```
[the remark this answers is not in the document]
Comment (EdgesFTW) [reply logpyle/runalyzer.py:48]: A double database close is a safe operation in python's sqlite3

File: .github/workflows/ci.yaml [hunk 8]
```diff
@@ -60,10 +60,21 @@ jobs:
             ## Check that t_log is within [0, 1]
             runalyzer -m summary.sqlite -c 'assert all([0 < r[0] < 1 for r in q("select $t_log.max")])'
 
+            ## Check that gathered and auto-gathered are equivilant for t_log non-mpi case
```
Comment (matthiasdiener) [inline .github/workflows/ci.yaml:63]:
```suggestion
            ## Check that gathered and auto-gathered are equivalent for t_log non-mpi case
```

File: doc/log.rst [hunk 9]
```diff
@@ -20,28 +20,32 @@ given as "watches" are also periodically logged to the console.
 To analyze this data, we need some tools that are distributed as part of this
 package.
 
-To prepare this data set for analysis, run ``runalyzer-gather`` as a script::
+You may use ``runalyzer`` to analyze the log file::
+
```
Comment (EdgesFTW) [inline doc/log.rst:24]: A message should be sent notifying the user that an in-memory db has been created

File: test/test_logManager.py [hunk 10]
```diff
@@ -317,6 +317,10 @@ def is_unique_filename(str: str):
         else:
             return False
 
+    files = [f for f in os.listdir() if is_unique_filename(f)]
+    for f in files:
+        os.remove(f)
+
```
Comment (matthiasdiener) [inline test/test_logManager.py:323]: Like we discussed, maybe remove this here

File: test/test_gather.py [hunk 11]
```diff
@@ -0,0 +1,222 @@
+import os
+import subprocess
+
+
+def test_auto_gather_single():
+    # run example
+    os.system("python ../examples/log.py")
```
Comment (matthiasdiener) [inline test/test_gather.py:7]: Can you run this test without `os.system`? Like just creating one or more `LogManager`s directly>

File: test/test_gather.py [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+    assert os.path.exists("log.sqlite"), "The logging file was not generated."
+
+    # check schema
+
+    # ensure logging table exists
+    result = subprocess.run("runalyzer log.sqlite -c 'db.print_cursor(\
```
Comment (matthiasdiener) [inline test/test_gather.py:13]: Same as above, can you run this by calling the gathering function dierctly?

File: .github/workflows/ci.yaml [hunk 13]
```diff
@@ -132,8 +171,9 @@ jobs:
 
         - name: Run tests
           run: |
+            sudo apt-get update && sudo apt-get install -y libopenmpi-dev
             pip install -e .
             cd test
-            python -m pip install pytest pymbolic psutil matplotlib
+            python -m pip install pytest pymbolic psutil matplotlib mpi4py
```
Comment (matthiasdiener) [inline .github/workflows/ci.yaml:177]: Is MPI required for pytest?

Comment (EdgesFTW) [reply .github/workflows/ci.yaml:177]: Not since the change to call into runalyzer libraries for testing gather

**Comment by matthiasdiener** [comment]
LGTM, thanks!

**Pull request merged**